### PR TITLE
[clang-format] Add DiagHandler parameter to format::getStyle()

### LINKED ISSUE #94205: Impossible to redirect diagnostic output getStyle function
We in QtCreator use getStyle function while the user typing the code. If the .clang-format file contains some issue then every call generates a stdout. It is clatter up the whole QtC output so to mitigate it would be good to have the possibility to redirect getStyle function diagnostic output. That possibility already exists in parseConfiguration function.

### PR DESCRIPTION
**Pull request by pointhex**
It allows to control of error output for the function.

Closes #94205.

### CODE REVIEW

**Comment by mydeveloperday** [comment]
is there a github issue for this?

**Comment by pointhex** [comment]
@mydeveloperday There is no issue, It is my request. I implemented it without creating an issue or suggestion. Should I?

**Comment by HazardyKnusperkeks** [comment]
For the record, I see no harm in this PR and would also merge it without an issue. Just giving @mydeveloperday a bit more time to answer. :)

**Comment by pointhex** [comment]
Hi @mydeveloperday, friendly reminder :)

**Comment by mydeveloperday** [comment]
Please log an issue explaining the problem you are trying solve, its not clear to me what problem led you to this.

**Comment by pointhex** [comment]
Done: https://github.com/llvm/llvm-project/issues/94205

**Review by HazardyKnusperkeks — APPROVED** [review]
Now that I see you want to incorporate it into QtCreator I'm even more in favor of adding it. ;)

**Review by mydeveloperday — COMMENTED** [review]
I don't have a major issue, but I think it needs a unit test otherwise we might forget how you are using it.

**Comment by mydeveloperday** [comment]
I also think we need @owenca  opinion

**Review by owenca — COMMENTED** [review]
Can you call `parseConfiguration()` instead of `getStyle()` in QtCreator while the user is entering code? It seems inefficient to call `getStyle()` all the time.

**Comment by pointhex** [comment]
> Can you call `parseConfiguration()` instead of `getStyle()` in QtCreator while the user is entering code? It seems inefficient to call `getStyle()` all the time.

I do it in the places where I can use parseConfiguration instead. But in some cases I need getStyle and it's also called quite often. I have one optimisation in my mind, but I would like to have that patch as well anyway.

Am I right that no one has a major issue with the patch? So fixing the "const" issue and adding a unit test will let that in?

File: clang/lib/Format/Format.cpp [hunk 1]
```diff
@@ -3945,20 +3945,23 @@ const char *DefaultFallbackStyle = "LLVM";
 
 llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>>
 loadAndParseConfigFile(StringRef ConfigFile, llvm::vfs::FileSystem *FS,
-                       FormatStyle *Style, bool AllowUnknownOptions) {
+                       FormatStyle *Style, bool AllowUnknownOptions,
+                       llvm::SourceMgr::DiagHandlerTy DiagHandler = nullptr) {
```
Comment (owenca) [inline clang/lib/Format/Format.cpp:3949]: It seems we don't need the default `nullptr`.

File: clang/lib/Format/Format.cpp [hunk 2]
```diff
@@ -4026,12 +4030,13 @@ Expected<FormatStyle> getStyle(StringRef StyleName, StringRef FileName,
   // Reset possible inheritance
   Style.InheritsParentConfig = false;
 
-  auto dropDiagnosticHandler = [](const llvm::SMDiagnostic &, void *) {};
+  auto diagHandlerOrDropHandling =
+      DiagHandler ? DiagHandler : [](llvm::SMDiagnostic const &, void *) {};
```
Comment (owenca) [inline clang/lib/Format/Format.cpp:4034]: It seems we can just delete line 4029. See below.

File: clang/lib/Format/Format.cpp [hunk 3]
```diff
@@ -4026,12 +4030,13 @@ Expected<FormatStyle> getStyle(StringRef StyleName, StringRef FileName,
...
 
   auto applyChildFormatTexts = [&](FormatStyle *Style) {
     for (const auto &MemBuf : llvm::reverse(ChildFormatTextToApply)) {
       auto EC = parseConfiguration(*MemBuf, Style, AllowUnknownOptions,
-                                   dropDiagnosticHandler);
+                                   diagHandlerOrDropHandling);
```
Comment (owenca) [inline clang/lib/Format/Format.cpp:4039]: It seems we only need to pass `DiagHandler`.

File: clang/lib/Format/Format.cpp (hunk 1 above)
Comment (pointhex) [reply clang/lib/Format/Format.cpp]: Done

File: clang/lib/Format/Format.cpp (hunk 2 above)
Comment (pointhex) [reply clang/lib/Format/Format.cpp]: Done

File: clang/lib/Format/Format.cpp (hunk 3 above)
Comment (pointhex) [reply clang/lib/Format/Format.cpp]: Done

**Comment by pointhex** [comment]
@mydeveloperday @owenca Kind reminder.

**Comment by owenca** [comment]
> Am I right that no one has a major issue with the patch? So fixing the "const" issue and adding a unit test will let that in?

Can you add a unit test?

**Comment by pointhex** [comment]
> Can you add a unit test?
Yes, no problem, I just didn't get if it is needed or not. Could you please tell me if you already have a unit test for that function and where? I didn't manage to find it.

**Comment by owenca** [comment]
@mydeveloperday requested it. See https://github.com/llvm/llvm-project/pull/91317#pullrequestreview-2095010527.

There are calls to `getStyle()` in `ConfigParseTest.cpp`. Maybe add a test there?

**Comment by pointhex** [comment]
I used to redirect output. I guess that is enough. Maybe, you know a better solution for that.
```    ::testing::internal::CaptureStderr();
    auto Style = getStyle("{invalid_key=invalid_value}", "a.h", "LLVM", "", &FS, false);
    ...
    const std::string output = ::testing::internal::GetCapturedStderr();
```

File: clang/unittests/Format/ConfigParseTest.cpp [hunk 4]
```diff
@@ -1452,6 +1452,35 @@ TEST(ConfigParseTest, GetStyleOfSpecificFile) {
...
+    // With output
+    ::testing::internal::CaptureStderr();
+    llvm::vfs::InMemoryFileSystem FS;
+    auto Style = getStyle("{invalid_key=invalid_value}", "a.h", "LLVM", "", &FS, false);
+
+    const std::string output = ::testing::internal::GetCapturedStderr();
+
+    ASSERT_FALSE((bool)Style);
+    ASSERT_FALSE(output.empty());
+    llvm::consumeError(Style.takeError());
```
Comment (owenca) [inline clang/unittests/Format/ConfigParseTest.cpp:1465]: IMO we don't need this test.

File: clang/unittests/Format/ConfigParseTest.cpp [hunk 5]
```diff
@@ -1452,6 +1452,35 @@ TEST(ConfigParseTest, GetStyleOfSpecificFile) {
...
+                           "",
+                           &FS,
+                           false,
+                           [](const llvm::SMDiagnostic &, void *) {});
+
+    const std::string output1 = ::testing::internal::GetCapturedStderr();
+
+    ASSERT_FALSE((bool)Style1);
+    ASSERT_TRUE(output1.empty());
+    llvm::consumeError(Style1.takeError());
```
Comment (owenca) [inline clang/unittests/Format/ConfigParseTest.cpp:1481]:
```suggestion
  // Suppress stderr.
  testing::internal::CaptureStderr();
  auto Style1 = getStyle("{invalid_key=invalid_value}", "a.h", "LLVM", "", &FS,
                         /*AllowUnknownOptions=*/true,
                         [](const llvm::SMDiagnostic &, void *) {});
  const auto output1 = testing::internal::GetCapturedStderr();
  ASSERT_TRUE((bool)Style1);
  ASSERT_TRUE(output1.empty());
```

File: clang/lib/Format/Format.cpp [hunk 6]
```diff
@@ -4027,12 +4031,10 @@ Expected<FormatStyle> getStyle(StringRef StyleName, StringRef FileName,
   // Reset possible inheritance
   Style.InheritsParentConfig = false;
 
-  auto dropDiagnosticHandler = [](const llvm::SMDiagnostic &, void *) {};
-
```
[the remark this answers is not in the document]
Comment (pointhex) [reply clang/lib/Format/Format.cpp]: No problem.

File: clang/lib/Format/Format.cpp [hunk 7]
```diff
@@ -4027,12 +4031,10 @@ Expected<FormatStyle> getStyle(StringRef StyleName, StringRef FileName,
...
   auto applyChildFormatTexts = [&](FormatStyle *Style) {
     for (const auto &MemBuf : llvm::reverse(ChildFormatTextToApply)) {
       auto EC = parseConfiguration(*MemBuf, Style, AllowUnknownOptions,
-                                   dropDiagnosticHandler);
+                                   DiagHandler);
```
[the remark this answers is not in the document]
Comment (pointhex) [reply clang/lib/Format/Format.cpp]: Done

File: clang/unittests/Format/ConfigParseTest.cpp (hunk 5 above)
Comment (pointhex) [reply clang/unittests/Format/ConfigParseTest.cpp]: Done

File: clang/unittests/Format/ConfigParseTest.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (pointhex) [reply clang/unittests/Format/ConfigParseTest.cpp]: Done

**Review by owenca — APPROVED** [review]

**Comment by pointhex** [comment]
@owenca  Thanks a lot!

**Pull request merged**